### DISCUSSION TITLE
Power simulation using global route only

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#6219] [General]
Is it possible to run power simulation using global route results to save build times or do I have to run all the way through detailed route?

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/6cb2a21d87ec898c346a82fe0785f6f1f08239d4/flow/designs/asap7/mock-array/config.mk#L54-L57

I wanted to try on mock-array, but immediately ran into trouble:

```
make DESIGN_CONFIG=designs/asap7/mock-array/Element/config.mk grt
cp results/asap7/mock-array_Element/base/5_1_grt.odb results/asap7/mock-array_Element/base/5_route.odb
cp results/asap7/mock-array_Element/base/5_1_grt.sdc results/asap7/mock-array_Element/base/5_route.sdc
make DESIGN_CONFIG=designs/asap7/mock-array/Element/config.mk do-final
```

This failed:

```
Running density_fill.tcl, stage 6_1_fill
exec cp ./results/asap7/mock-array_Element/base/5_route.odb ./results/asap7/mock-array_Element/base/6_1_fill.odb
Elapsed time: 0:01.35[h:]min:sec. CPU time: user 1.23 sys 0.17 (103%). Peak memory: 213888KB.
cp ./results/asap7/mock-array_Element/base/5_route.sdc ./results/asap7/mock-array_Element/base/6_1_fill.sdc
cp ./results/asap7/mock-array_Element/base/5_route.sdc ./results/asap7/mock-array_Element/base/6_final.sdc
Running final_report.tcl, stage 6_report
Deleted 0 routing obstructions
[INFO RCX-0431] Defined process_corner X with ext_model_index 0
[INFO RCX-0029] Defined extraction corner X
[INFO RCX-0435] Reading extraction model file /home/oyvind/OpenROAD-flow-scripts/flow/platforms/asap7/rcx_patterns.rules ...
[INFO RCX-0436] RC segment generation Element (max_merge_res 50.0) ...
[INFO RCX-0040] Final 0 rc segments
[INFO RCX-0439] Coupling Cap extraction Element ...
[INFO RCX-0440] Coupling threshhold is 0.1000 fF, coupling capacitance less than 0.1000 fF will be grounded.
[INFO RCX-0442] 98% of 2484 wires extracted
[INFO RCX-0442] 100% of 2484 wires extracted
[WARNING RCX-0107] Nothing is extracted out of 1843 nets!
[INFO RCX-0134] Can't execute write_spef command. There's no extraction data.
Error: final_report.tcl, 62 cannot read file ./results/asap7/mock-array_Element/base/6_final.spef.
Command exited with non-zero status 1
Elapsed time: 0:01.40[h:]min:sec. CPU time: user 1.24 sys 0.16 (100%). Peak memory: 222720KB.
make[1]: *** [Makefile:841: do-6_report] Error 1
make: *** [Makefile:853: do-finish] Error 2
```

**Comment by oharboe** [comment]
Running only through iteration 0 in detailed route, I get some sort of result...

Still: even iteration 0 & 1 can take quite a while on big designs... I wanted to just use the result after global route in CI to ensure that nothing breaks.

```
make DESIGN_CONFIG=designs/asap7/mock-array/config.mk DETAILED_ROUTE_END_ITERATION=0 nuke final simulate power
```

**Comment by maliberty** [comment]
It is possible in principle to run it after global routing though ORFS is not configured to do so.  You wouldn't be reading a spef in that case but would use `estimate_parasitic -global_routing`.  The parasitics will differ somewhat from a fully routed design so the power drawn by the instances will vary as well.  You will likely get a decent result but nobody has tried afaik.

**Reply by oharboe** [reply]
Silly question: how do I use estimate_parasitics, an openroad command, when it is OpenSTA that must run the power.tcl script?

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/0b5c61d25a019677b4e93a65c84cd710cfd47589/flow/designs/asap7/mock-array/config.mk#L57

**Reply by maliberty** [reply]
In that case you can't.

**Reply by oharboe** [reply]
I see. At least then I know that we have to run some detailed routing, it's the cost of doing business w.r.t. this sort of power simulation.

Any thoughts on what inaccuracies to expect with DRC errors?

I.e. if I run only 0 or 1 iterations, will I get some inaccuracies or just completely wrong numbers?

```
make DESIGN_CONFIG=designs/asap7/mock-array/config.mk DETAILED_ROUTE_END_ITERATION=0 nuke final simulate power
```

**Reply by maliberty** [reply]
Probably some inaccuracy.  I just remembered that we added `-spef_file filename` to estimate_parasitics so you could write parasitics from global routing.

**Reply by oharboe** [reply]
With https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/2587, the below should work?



```
make DESIGN_CONFIG=designs/asap7/mock-array/Element/config.mk grt
make ABSTRACT_SOURCE=5_1_grt DESIGN_CONFIG=designs/asap7/mock-array/Element/config.mk do-generate_abstract
cp results/asap7/mock-array_Element/base/5_1_grt.v results/asap7/mock-array_Element/base/6_final.v
cp results/asap7/mock-array_Element/base/5_1_grt_default.spef results/asap7/mock-array_Element/base/6_final.spef
make ABSTRACT_SOURCE=5_1_grt DESIGN_CONFIG=designs/asap7/mock-array/config.mk grt
cp results/asap7/mock-array/base/5_1_grt.v results/asap7/mock-array/base/6_final.v
cp results/asap7/mock-array/base/5_1_grt_default.spef results/asap7/mock-array/base/6_final.spef
cp results/asap7/mock-array/base/5_1_grt.sdc results/asap7/mock-array/base/6_final.sdc
make DESIGN_CONFIG=designs/asap7/mock-array/config.mk simulate power
```

**Reply by maliberty** [reply]
Yes in theory

**Reply by oharboe** [reply]
Some sort of mismatch in .v and .spef file... Issue filed: https://github.com/The-OpenROAD-Project/OpenROAD/issues/6225

**Comment by oharboe** [comment]
The question in this thread was answered, but the issue with getting power simulation to work based on global routing needs to be resolved.

https://github.com/The-OpenROAD-Project/OpenROAD/issues/6225
